### ISSUE TITLE
Cant work with multiple texture in fragment shader

### ISSUE THREAD

**Issue by BarisUckardes** [mellinoe/veldrid#315]
Having a difficult time when i try use multiple resources.

Now, i have three images

->Pacman image (First Image)
->Flower image (Second Image)
->Brick image (Third Image)

i want to overlay and mix them in my fragment shader, but when i bind multiple textures it always falls back to the first image i bind.

For example 

+If i only output the color of the pacman image it outputs correctly
+if i sample pacman image and then output flower image, it outpus correctly
+if i sample both pacman and flower image output brick image,it outputs correctly

BUT !

+if i dont sample pacman and try to output flower or brick texture, output color always fallsback to pacman image

This is my fragment shader code

` 
private const string FragmentCode = @"
#version 450

                                            layout(location = 0) in vec2 tex_coords;
                                            layout(location = 0) out vec4 fsout_Color;
                                            
                                            layout(set = 1,binding = 1) uniform texture2D pacManTexture;
                                            layout(set = 1,binding = 2) uniform sampler pacManSampler;

                                            layout(set = 2,binding = 1) uniform texture2D flowerTexture;
                                            layout(set = 2,binding = 2) uniform sampler flowerSampler;
                                            
                                            layout(set = 3,binding = 1) uniform texture2D brickTexture;
                                            layout(set = 3,binding = 2) uniform sampler brickSampler;

                                            void main()
                                            {
                                                //fsout_Color =texture(sampler2D(pacManTexture,pacManSampler),tex_coords);
                                                //fsout_Color = texture(sampler2D(flowerTexture,flowerSampler),tex_coords);
                                                fsout_Color =texture(sampler2D(brickTexture,brickSampler),tex_coords);

                                            }";
`
This image outpus the first image which is the first resource set that i bind to fragment shader stage.

This is how i create my texture,textureView,resourceLayout and resourceSets

![image](https://user-images.githubusercontent.com/68399075/97576116-cd332200-19fe-11eb-9fc9-7e97152db380.png)

I must be missing something vital here or is this some kind of bug.

My guess is doing something wrong with resource sets or layout set&binding numbers in the fragment shader

BTW I tried both opengl and dx11 backend i did no difference

**Comment by mellinoe** [comment]
The problem is that the code in your shader is simply overwriting the color of each pixel with the last texture that you sample. That code isn't going to blend them or mix them in any way.

What you likely want to do is issue 3 separate draw calls, each one using a different texture. With the appropriate blend settings (e.g. using `BlendStateDescription.SingleAlphaBlend`), the three draw calls will blend together based on each texture's transparency, and the order in which you draw things.

**Comment by BarisUckardes** [comment]
Well the way you describe seems a bit inefficient? draw call per texture.I still dont understand the overwriting mechanic.It always outputs the color of the first texture resource set i bind.I mean in your offscreen example mirror fragment shader sampled two textures at the same time.(In my case the first texture will be displayed in my scenario)

By the way my shader never outputs the last pixel, there is no overriding problem.It always outputs the first resource set i bind, in this case is the pacman.

I set resource sets as pacman,flower and then the brick
when try to sample and output the brick texture it always shows the pacman not the brick.
if i try to sample and output flower it still shows the pacman.

if i change the resource set order to flower,brick and pacman even if i sample the pacman it shows the flower.

I tried to make single resource set as you did in the offscreen example it didnt worked.

**Comment by mellinoe** [comment]
Can you share a minimal repro that I can look at, or at least the code involved with binding the resource sets here? Without seeing that code, it's hard to say what's happening here.

**Comment by BarisUckardes** [comment]
Well i can share the codes involed resourceLayout/resourceSet creation and bindings.If you need further resource i can share the whole project.

This is where i load my texture from file using sharp image
+Im using my own texture wrap class(RTexture2D) to contain resourceLayouts and resourceSets
+I even set the resourceLayout&resourceSet Name property so i can debug which layout binds and where and all seems in correct place

![image](https://user-images.githubusercontent.com/68399075/97690547-2c03a480-1aae-11eb-96cf-3710f0f65c6a.png)

This is how i invoke this method
![image](https://user-images.githubusercontent.com/68399075/97690803-8866c400-1aae-11eb-8c68-68500dc2c53f.png)

And this is how my engine's InternalPassRendering method uses resourceLayouts and resourceSets
This RUniformBuffer structure holds my uniform buffer(ModelViewProjection)
It contains
+ResourceLayouts
+ResourceSets
![image](https://user-images.githubusercontent.com/68399075/97691087-eeebe200-1aae-11eb-916b-38d2d2b7eab0.png)

I set my vertex&index buffers at the begining of the render call
![image](https://user-images.githubusercontent.com/68399075/97691235-265a8e80-1aaf-11eb-94dc-76b940e1d89d.png)

Creating pipeline description and assigning my uniform buffer's resourcelayout at first
![image](https://user-images.githubusercontent.com/68399075/97691344-51dd7900-1aaf-11eb-99c4-1da93a19ebaa.png)

Here i get all my texture resources(pacman,flower,brick) from my material structure
materialResourceLayout list and materialResourceSet list contains the correct order of textures (pacman,flower,brick) I CHECKED
![image](https://user-images.githubusercontent.com/68399075/97691418-63bf1c00-1aaf-11eb-986c-893815b42674.png)

I simply merge my resource layout array
Now pipeline description's resource layout array contains
+Uniform Buffer(MVP)
+Pacman Texture
+Flower Texture
+Brick Texture
I CHECKED
![image](https://user-images.githubusercontent.com/68399075/97691561-910bca00-1aaf-11eb-8a08-42864da3aff6.png)

I set my shader set as well as input layout
![image](https://user-images.githubusercontent.com/68399075/97691682-b8629700-1aaf-11eb-88b7-b23ee1b075d1.png)

Setting my pipeline output,creating my pipeline and setting the new pipeline
![image](https://user-images.githubusercontent.com/68399075/97691743-d0d2b180-1aaf-11eb-8f60-f3179b337494.png)

Here i set my uniform buffer(MVP) from the start
![image](https://user-images.githubusercontent.com/68399075/97691783-dcbe7380-1aaf-11eb-9706-9582141ab695.png)

And then setting all the texture resources i had
![image](https://user-images.githubusercontent.com/68399075/97691859-fd86c900-1aaf-11eb-98e8-57e29c411c9e.png)

My drawcall
![image](https://user-images.githubusercontent.com/68399075/97691903-0e373f00-1ab0-11eb-93c9-024431da1e92.png)

**Comment by deryaza** [comment]
I belive what you need to do is to multiply the fsout_Color, not reassign it.

**Comment by BarisUckardes** [comment]
> I belive what you need to do is to multiply the fsout_Color, not reassign it.

You are right when we try to mix colors but the problem is not about the HOW TO MIX 2 textures.Why i cant sample the textures i want.
Lets think of an example where i set an float value a threshold
if(threshold > 0.8)
return pacman color;
else if(threshod > 0.6 )
return flower color;
else
return brick color

in this case it always returns the first texture i bind to the pipeline.(Which is the pacman texture)

**Comment by deryaza** [comment]
I still dont get it. Do you want one object (like on screen) to change its texture to be ether pacman, flower or brick?
Also, I think your texture binding is wrong. Each set should strat from zero.

**Comment by BarisUckardes** [comment]
if you want to switch a texture according to a value, you need to sample all the textures if you want output the one you want.
Another example you want a height fog shader.
if the height is above 50 meters you want to use texture A, if its not you want to use texture B.
In my problem you cannot use the second texture unless you first sample the initial one.

**Comment by BarisUckardes** [comment]
I made an another test

Sampling and storing the color value does not help
![image](https://user-images.githubusercontent.com/68399075/97808058-74b98a00-1c75-11eb-8a6f-8f73e825b645.png)

Fragment Shader above still outputs the first texture that i bind to the pipeline

Like so
![image](https://user-images.githubusercontent.com/68399075/97808080-ac283680-1c75-11eb-8015-9661f5386666.png)

**Comment by mellinoe** [comment]
Can you share a runnable project that reproduces the issue? Feel free to share privately if you need. Running this under RenderDoc would make it very obvious what the problem is -- most likely a resource binding issue. Most of what you screenshotted above looks reasonable, but there are still things hidden behind other calls that could be problematic. If I can run the code, then I can get to the bottom of it much more quickly.

**Comment by BarisUckardes** [comment]
Here all the source code and the executable

http://www.mediafire.com/file/ch40eo7kw32zssx/REngine.7z/file

If you need a peek, you can start with program entry point class(REngineEntryPoint).

**Comment by mellinoe** [comment]
There's a couple of things going on here. Your shader is written in a way that leaves 2/3 of the textures unused. "Unused" in this context includes optimizations, so unless the values sampled from the textures are actually used in some way to affect the output, they don't count. In this shader, the values sampled from the texture are immediately discarded -- they aren't actually used for anything. Since the resources don't exist in the compiled form that gets passed into Veldrid, the textures end up bound to the wrong slots, which is why it seems like you are not sampling from the resource you expect.

If your shader was written to use all of the textures (by blending them, etc. as suggested above), then the code will work as expected. Alternatively, if you truly only need to sample from one texture, and modify your shader code to reflect that (by removing the unused resources), then it would also work as expected.

To summarize, if your shader only samples from one texture, then remove the others from the code. If you need to sample from multiple textures conditionally like this:

>Lets think of an example where i set a float value, a threshold
> ```
> if(threshold > 0.8)
> return pacman color;
> else if(threshod > 0.6 )
> return flower color;
> else
> return brick color
> ```

then that would work fine, because all of the resources are actually used in the shader.

Also note that you could technically use some different options with Veldrid.SPIRV to get your shader working as-is (basically by turning off optimizations), but that is a little bit less user-friendly.

**Comment by BarisUckardes** [comment]
So its basically a SPIRV optimization thing.IF i turn off the optimization what kind of impact i should expect in terms of memory bandwith and fillrate or any other extra overhead?.

By the way i looked up a little bit about Veldrid.SPIRV optimizations, but i couldnt find any kind of options or enums when compiling the shaders.Only thing i found is the CrossCompilerOptimizations struct but it didnt looked like it was containing the option i need.So how/where do i disable these optimizations.

**Comment by mellinoe** [comment]
> So its basically a SPIRV optimization thing.IF i turn off the optimization what kind of impact i should expect in terms of memory bandwith and fillrate or any other extra overhead?.

It's hard to say. The shaders that come out of Veldrid.SPIRV will definitely be more complicated / less optimized than normal, but those are just the HLSL/GLSL text shaders that go into the Direct3D and OpenGL compilers, which likely apply some optimizations themselves before they are used. My guess is that there would be a minor difference. You wouldn't want to disable optimizations if you were targeting Vulkan, because it doesn't experience this problem (all binding points are baked into the SPIR-V in a way that Veldrid natively understands).

> By the way i looked up a little bit about Veldrid.SPIRV optimizations, but i couldnt find any kind of options or enums when compiling the shaders.Only thing i found is the CrossCompilerOptimizations struct but it didnt looked like it was containing the option i need.So how/where do i disable these optimizations.

The key is to call `CompileGlslToSpirv` directly, and pass a `GlslCompileOptions` with Debug=true. This basically prevents optimizations from being applied. I changed your CompileShaderSet function to the following to fix the rendering:

```C#
public Shader[] CompileShaderSet(ShaderDescription vertexDescription, ShaderDescription pixelDescription)
{
    var pixelSpirvCompilation = SpirvCompilation.CompileGlslToSpirv(
        Encoding.UTF8.GetString(pixelDescription.ShaderBytes), "FS", ShaderStages.Fragment, new GlslCompileOptions(debug: true));
    pixelDescription.ShaderBytes = pixelSpirvCompilation.SpirvBytes;

    return TargetGraphicsFactory.CreateFromSpirv(vertexDescription, pixelDescription);
}
```

**Comment by BarisUckardes** [comment]
> The key is to call CompileGlslToSpirv directly, and pass a GlslCompileOptions with Debug=true. This basically prevents optimizations from being applied. I changed your CompileShaderSet function to the following to fix the rendering:

This works perfectly.

Eric thank you for your consistent support.This was my first issue which i posted on github.Do you have any critics about the way i post my issue, i would like to know how things done here.

Other than that it seems this issue is officially solved and its ready to be closed.

**Comment by mellinoe** [comment]
I'm glad I was able to help @BarisUckardes . In terms of communicating your problem and getting to the solution the fastest and easiest, nothing beats providing a minimal reproducible example that I can compile and run myself. The screenshots you posted above were helpful, but not quite enough to get to the bottom of the problem. Once I can browse the code and run it, it's a lot easier to provide help.
